hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/ArchivedTimelineWriter.java

```diff
@@ -0,0 +1,382 @@
+    }
+
+    // cleaning
+    clean(context, layer);
+  }
+
+  private Option<String> doCompact(HoodieArchivedManifest manifest, int layer) throws IOException {
+    // 1. list all the files that belong to current layer
+    List<HoodieArchivedManifest.FileEntry> files = manifest.getFiles()
+        .stream().filter(file -> HoodieArchivedTimeline.isFileFromLayer(file.getFileName(), layer)).collect(Collectors.toList());
+
+    int compactionBatchSize = config.getArchiveMergeFilesBatchSize();
+
+    if (files.size() >= compactionBatchSize) {
+      // 2. sort files by min instant time (implies ascending chronological order)
+      files.sort(HoodieArchivedManifest.FileEntry::compareTo);
+      List<String> candidateFiles = getCandidateFiles(files, compactionBatchSize);
+      if (candidateFiles.size() < 2) {
+        // the file is too large to compact, returns early.
+        return Option.empty();
+      }
+      String compactedFileName = compactedFileName(candidateFiles);
+
+      // 3. compaction
+      compactArchiveFiles(candidateFiles, compactedFileName);
+      // 4. update the manifest file
+      updateManifest(candidateFiles, compactedFileName);
+      LOG.info("Finishes compaction of archive files: " + candidateFiles);
+      return Option.of(compactedFileName);
+    }
+    return Option.empty();
+  }
+
+  public void compactArchiveFiles(List<String> candidateFiles, String compactedFileName) {
+    LOG.info("Starting to merge small archive files.");
+    try (HoodieFileWriter writer = openWriter(new Path(metaClient.getArchivePath(), compactedFileName))) {
+      for (String fileName : candidateFiles) {
+        // Read the archived file
+        try (HoodieAvroParquetReader reader = (HoodieAvroParquetReader) HoodieFileReaderFactory.getReaderFactory(HoodieRecord.HoodieRecordType.AVRO)
+            .getFileReader(metaClient.getHadoopConf(), new Path(metaClient.getArchivePath(), fileName))) {
+          // Read the meta entry
+          try (ClosableIterator<IndexedRecord> iterator = reader.getIndexedRecordIterator(HoodieArchivedInstant.getClassSchema(), HoodieArchivedInstant.getClassSchema())) {
+            while (iterator.hasNext()) {
+              IndexedRecord record = iterator.next();
+              writer.write(record.get(0).toString(), new HoodieAvroIndexedRecord(record), HoodieArchivedInstant.getClassSchema());
+            }
+          }
+        }
+      }
+    } catch (Exception e) {
+      throw new HoodieCommitException("Failed to merge small archive files", e);
+    }
+    LOG.info("Success to merge small archive files.");
+  }
+
+  /**
+   * Checks whether there is any unfinished compaction operation.
+   *
+   * @param context HoodieEngineContext used for parallelize to delete small archive files if necessary.
+   */
+  public void clean(HoodieEngineContext context, int compactedVersions) throws IOException {
+    // if there are more than 3 version of snapshots, clean the oldest files.
```

Comment:
how do we cleanup the partially written files, (i.e like the marker mechanism we have) 

- we wrote files (write or compaction) but did not produce manifest. 
- we wrote a manifest, but crashed before updating version. .. 

My take is to add a new method, that can scan the entire folder and delete any file that is not pointed to by active manifest/versions. We can trigger such cleanup once a day or sth? (or just leave it as a CLI command and people can do it if they wish?)

Reply:
Is this clean, we already list all the data files first, and diff that with the valid version files in our latest snapshots, all those invalid/partial written files should be removed.

We clean the manifest file in the similiar way.